File: add-two-numbers/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1:ListNode, l2: ListNode) -> ListNode:
        if l1 is None: #Handle case of empty lists
            return l2
        if l2 is None:
            return l1
        carry = 0
        head = l1
        p1 = l1
        p2 = l2
        prev = None
        while (p1 is not None and p2 is not None): #Both lists have remaining elements
            val1 = p1.val
            val2 = p2.val
            total = val1 + val2 + carry
            carry = total // 10 #Determine new carry
            total = total % 10 #Determine node value
            p1.val = total #Assign value to node

            prev = p1 #Keep previous node in case next is None
            p1 = p1.next
            p2 = p2.next
        while (p1 is not None and carry != 0): #l2 is empty
            total = p1.val + carry
            carry = total // 10
            total = total % 10
            p1.val = total
            prev = p1
            p1 = p1.next
        while (p2 is not None): #l1 is empty
            total = p2.val + carry
            carry = total // 10
            total = total % 10
            prev.next = ListNode(total)
            prev = prev.next
            p2 = p2.next
        if carry != 0: #l1 is empty, l2 is empty, carry is greater than 0
            prev.next = ListNode(carry)
        return head

File: add-two-numbers/add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution 
{
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) 
    {
        // Create a dummy node and pointers to iterate the lists
        // Create a carry for summations resulting in values > 9
        ListNode dummy = new ListNode(-1);
        ListNode p1 = l1, p2 = l2, current = dummy;
        int carry = 0;
        
        while (p1 != null || p2 != null)
        {
            // Extract values and perform basic addition
            int l1Val = p1 != null ? p1.val : 0;
            int l2Val = p2 != null ? p2.val : 0;
            int val = l1Val + l2Val + carry;
            
            // Determine if there is a new carry
            carry = val > 9 ? 1 : 0;
            
            // Create new node for the value and move ahead to it
            current.next = new ListNode(val % 10);
            current = current.next;
            
            // Move forward the listnode pointers for next addition
            if (p1 != null) p1 = p1.next;
            if (p2 != null) p2 = p2.next;
        }
        
        // Check if there was a carry on final addition
        if (carry > 0)
            current.next = new ListNode(carry);
        
        return dummy.next;
    }
}